torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp

```diff
@@ -1120,45 +1121,103 @@ bool isSortableTupleType(
   return true;
 }
 
-bool isSortableListOfObjectsOrTuples(
+bool isSameTupleOptionalType(const TupleTypePtr& t1, const TupleTypePtr& t2) {
+  auto types1 = t1->elements();
+  auto types2 = t2->elements();
+  size_t n1 = types1.size();
+  size_t n2 = types2.size();
+
+  if (n1 != n2)
+    return false;
+  for (size_t i = 0; i < n1; i++) {
+    auto cur_t1 = types1[i];
+    auto cur_t2 = types2[i];
+
+    if (cur_t1->kind() == TypeKind::NoneType ||
+        cur_t2->kind() == TypeKind::NoneType) {
+      continue;
+    } else if (*cur_t1 != *cur_t2) {
+      return false;
+    }
+  }
+  return true;
+}
+
+// Check whether ivalues are sortable. If so,
+// update comparator if we can get one.
+bool checkAndGetComparator(
     c10::List<IValue>& ivalues,
+    bool reverse,
+    c10::IValueComparator& comparator,
     std::stringstream& why_not) {
   if (ivalues.empty()) {
     return true;
   }
 
+  // Find first type that is not None.
   auto type = ivalues.get(0).type();
+  const size_t n = ivalues.size();
+  size_t i = 0;
+  while (type == NoneType::get()) {
+    if (i == n - 1) {
+      // Everything is None. No need to get a comparator.
+      return true;
+    }
+    i++;
+    type = ivalues.get(i).type();
+  }
+
+  auto tuple_type = type->cast<TupleType>();
   // We assume lists have homogenous types, use first element to determine
   // best sorting methods. If in the future we need to support heterogenous
   // types inside list, then sorting needs to have runtime sortable checks.
-  const size_t n = ivalues.size();
-  for (size_t i = 0; i < n; ++i) {
-    const IValue& v = ivalues.get(i);
+  for (size_t j = i + 1; j < n; ++j) {
+    const IValue& v = ivalues.get(j);
     auto curr_type = v.type();
-    if (*curr_type != *type) {
+    if (*curr_type != *type && curr_type->kind() != TypeKind::NoneType) {
+      // Give it another chance to see if they can be viewed as
+      // Tuple[Optional[x]]
+      auto curr_tuple_type = curr_type->cast<TupleType>();
+      if (tuple_type && curr_tuple_type &&
+          isSameTupleOptionalType(tuple_type, curr_tuple_type)) {
```

Comment:
I think you will run into problems here with a test case like this: `[(None, 3), (3, 4), ("hello", 4)]`. `isSameTupleOptionalType` will find that `(None, 3)` and `(3, 4)` are the same optional tuple type (`Tuple[Optional[int], Optional[int]]`), and that `(None, 3)` and `("hello", 4)` are the same optional tuple type (`Tuple[Optional[str], Optional[int]]`), but `(3, 4)` and `("hello", 4)` have different types and cannot be compared.